### PR TITLE
HPCC-31545 Thread pool wait time reporting

### PR DESCRIPTION
**Pull request by mckellyln**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jthread.cpp [hunk 1]
```diff
@@ -1017,13 +1026,11 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
                         throw MakeStringException(0, "No threads available in pool %s", poolname.get());
                     IWARNLOG("Pool limit exceeded for %s", poolname.get());
```
Comment (mckellyln) [inline system/jlib/jthread.cpp:1027]: This code continues on with a new thread even when it timed out waiting for a free slot (when timeout == 0)
Is that what we want ?  Should it instead wait 'forever' for a free slot when timeout == 0 ?

File: system/jlib/jthread.cpp [hunk 2]
```diff
@@ -1017,13 +1026,11 @@ class CThreadPool: public CThreadPoolBase, implements IThreadPool, public CInter
...
                 }
-                else
-                    timedout = false;
             }
             if (traceStartDelayPeriod)
             {
                 ++startsInPeriod;
-                if (timedout)
+                if (waited)
```
Comment (jakesmith) [inline system/jlib/jthread.cpp:1033]: Don't think it really matters, but if there's an accumulate delay, but as it crosses the 1 minute mark (the 'traceStartDelayPeriod' period), 'waited' isn't true, then it won't report until the next delayed thread.

Perhaps change to:
```
            if (traceStartDelayPeriod)
            {
                ++startsInPeriod;
                if (waited)
                    startDelayInPeriod += startTimer.elapsedCycles();
                if (startDelayInPeriod && (overAllTimer.elapsedCycles() >= queryOneSecCycles()*traceStartDelayPeriod)) // check avg. delay per minute
                {
....
                }
            }
```

File: thorlcr/slave/slavmain.cpp [hunk 3]
```diff
@@ -1185,7 +1185,7 @@ class CKJService : public CSimpleInterfaceOf<IKJService>, implements IThreaded,
     {
         Owned<CProcessorFactory> factory = new CProcessorFactory(*this);
         processorPool.setown(createThreadPool("KJService processor pool", factory, this, keyLookupMaxProcessThreads, 10000));
-        processorPool->setStartDelayTracing(60000);
+        processorPool->setStartDelayTracing(60);
```
Comment (jakesmith) [inline thorlcr/slave/slavmain.cpp:1188]: whoops, good change!
Perhaps we should just default it to 60seconds in CThreadPool ?

**Review by jakesmith — APPROVED** [review]
@mckellyln - looks good, can be merged as is. A couple of minor comments.

File: system/jlib/jthread.cpp (hunk 2 above)
Comment (mckellyln) [reply system/jlib/jthread.cpp:1033]: I agree, good find.  I would even say tho I want to know how many threads started even if there is no accumulated delay, so I don't think we need the extra check of if(startDelayInPeriod) to log the info.  See commit 2.

File: thorlcr/slave/slavmain.cpp (hunk 3 above)
Comment (mckellyln) [reply thorlcr/slave/slavmain.cpp:1188]: Sure if you want I can change default.

**Comment by mckellyln** [comment]
@jakesmith made those changes and am requesting a quick re-review, thanks.

**Review by ghalliday — APPROVED** [review]
Changes look good to me

**Review by jakesmith — APPROVED** [review]
@mckellyln - looks good. Please squash.

**Comment by mckellyln** [comment]
Squashed.

**Pull request merged**